trunk/src/modules/upscale.F90

```diff
@@ -18,83 +20,131 @@ module upscale
-      real(8), allocatable :: VERTC(:,:)
+      real(8), pointer :: VERTC(:,:)
 !
 !  ...number of elements required by visualization level (0 - 3)
       integer :: NR_ELEM
 !
 !  ...list of elements' vertices (indexing starts at 0)
-      integer, allocatable :: ELEM(:,:)
+      integer, pointer :: ELEM(:,:)
+!
    endtype vis
 !
    type(vis) :: TETR_VIS, PRIS_VIS, HEXA_VIS
-!
+
+
 contains
-!
+
+
 !> @brief return cell type (XDMF 2) for vis object
 !> @param[in] Etype - Element type
 !> @date  Feb 2023
    integer function ivis_type(Etype)
+!
       integer :: Etype
-      select case(Etype)
+!
+      if (SECOND_ORDER_VIS) then
+         select case(Etype)
+         case(TETR,MDLN); ivis_type = 38
+         case(PRIS,MDLP); ivis_type = 41
+         case(BRIC,MDLB); ivis_type = 50
+         end select
+      else
+         select case(Etype)
          case(TETR,MDLN); ivis_type = 6
          case(PRIS,MDLP); ivis_type = 8
          case(BRIC,MDLB); ivis_type = 9
-      end select
-   end function ivis_type
+         end select
+      endif
 !
+   end function ivis_type
+
+
+
 !> @brief return number of points to describe vis object
 !> @param[in] Etype - Element type
 !> @date Feb 2023
    integer function nobj_conf(Etype)
+!
       integer :: Etype
-      select case(Etype)
-         case(TETR,MDLN); nobj_conf = 4
-         case(PRIS,MDLP); nobj_conf = 6
-         case(BRIC,MDLB); nobj_conf = 8
-         case default
-            write(*,*) 'nobj_conf: unexpected type. stop.'
-            stop
-      end select
-   end function nobj_conf
 !
+      if (SECOND_ORDER_VIS) then
+         select case(Etype)
+            case(TETR,MDLN); nobj_conf = 10
+            case(PRIS,MDLP); nobj_conf = 18
+            case(BRIC,MDLB); nobj_conf = 27
+            case default
+               write(*,*) 'nobj_conf: unexpected type. stop.'
+               stop
+         end select
+      else
+         select case(Etype)
+            case(TETR,MDLN); nobj_conf = 4
+            case(PRIS,MDLP); nobj_conf = 6
+            case(BRIC,MDLB); nobj_conf = 8
+            case default
+               write(*,*) 'nobj_conf: unexpected type. stop.'
+               stop
+         end select
+      endif
+!
+   end function nobj_conf
+
+
+
 !> @brief return preloaded vis element configuration
 !> @param[in] Etype - Element type
 !> @date Feb 2023
    type(vis) function vis_on_type(Etype)
+!
       integer :: Etype
+!
       select case(Etype)
          case(TETR,MDLN); vis_on_type = TETR_VIS
          case(PRIS,MDLP); vis_on_type = PRIS_VIS
          case(BRIC,MDLB); vis_on_type = HEXA_VIS
       end select
-   end function vis_on_type
 !
+   end function vis_on_type
+
+
+
 !> @brief deallocate vis object (configuration)
 !> @param[in,out] V
 !> @date Feb 2023
    subroutine clear_vis(V)
+!
       type(vis), intent(inout) :: V
+!
       V%NR_VERT = 0
-      if (allocated(V%VERTC)) deallocate(V%VERTC)
+      if (associated(V%VERTC)) deallocate(V%VERTC)
```

Comment:
Now that `V%VERTC` and `V%ELEM` have changed to pointers, they must be nullified when the type(vis) objects are created because the initial association status of a pointer is undefined. Otherwise, this statement here doesn't have a well-defined behavior.

Reply:
The behavior of `associated` is not defined when the pointer has been initialized but not nullified or allocated. So the statement `if (associated(...)` isn't necessarily going to be either `true` or `false`

Reply:
But if it's `true` wouldn't `deallocate` do the nullification?

I didn't know we could default initialize to null using `=> null()` (still learning about Fortran pointers). The problem when I did this was that we needed to nullify the first time this was called; then do this deallocation the following times and I didn't see a good way to distinguish. I'll add that default initialization

Reply:
Default initialization worked; I updated that.

Reply:
Yes, this works (`null()` requires Fortran95 or later but can be used directly in initialization). The standard way is using the `nullify(ptr)` function but that would've to be called after creating a type object.